src/libraries/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.Win32.cs

```diff
@@ -245,95 +248,108 @@ private void ReadDirectoryChangesCallback(uint errorCode, uint numBytes, AsyncRe
             }
         }
 
-        private unsafe void ParseEventBufferAndNotifyForEach(byte[] buffer)
+        private unsafe void ParseEventBufferAndNotifyForEach(ReadOnlySpan<byte> buffer)
         {
-            Debug.Assert(buffer != null);
-            Debug.Assert(buffer.Length > 0);
+            ReadOnlySpan<char> oldName = ReadOnlySpan<char>.Empty;
 
-            fixed (byte* b = buffer)
+            // Parse each event from the buffer and notify appropriate delegates
+            for (;;)
             {
-                Interop.Kernel32.FILE_NOTIFY_INFORMATION* info = (Interop.Kernel32.FILE_NOTIFY_INFORMATION*)b;
+                // Validate the data we received in case it's corrupted. This can happen
+                // if we are watching files over the network with a poor connection.
 
-                ReadOnlySpan<char> oldName = ReadOnlySpan<char>.Empty;
+                // Verify the info object is within the expected bounds
+                if (sizeof(FILE_NOTIFY_INFORMATION) > (uint)buffer.Length)
+                {
+                    break;
+                }
+                ref readonly FILE_NOTIFY_INFORMATION info = ref MemoryMarshal.AsRef<FILE_NOTIFY_INFORMATION>(buffer);
```

Comment:
Even through MemoryMarshal.AsRef API is unsafe, it still provides checks for buffer bounds and enforces that you can only get the reference to the data after the buffer bounds have been checked.